keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1107 @@
+-->
+
+The proposal only adds new informantional data to the CRI API between kubelet
+and the container runtime with no user-visible changes which mitigates possible
+risks considerably.
+
+Data duplication/inconsistency with native resources could be considered a risk
+as those are passed down to CRI both as "raw" requests and limits and as
+"translated" resource control parameters (like cpu shares oom scoring etc). But
+this should be largely mitigated by code reviews and unit tests.
+
+## Design Details
+
+The proposal is that kubelet discloses full resources information from the
+PodSpec to the container runtime. This is accomplished by extending the
+ContainerConfig, UpdateContainerResourcesRequest and PodSandboxConfig messages
+of the CRI API.
+
+With this information, the runtime can for example do detailed resource
+allocation so that CPU, memory and other resources for each container are
+optimally aligned.
+
+The resource information is included in PodSandboxConfig so that the runtime
+can see the full picture of Pod's resource usage at Pod creation time, for
+example enabling more holistic resource allocation and thus better
+interoperability between containers inside the Pod.
+
+### CRI API
+
+#### PodSandboxConfig
+
+The PodSandboxConfig message (part of the RunPodSandbox request) will be
+extended to contain information about resources of all its containers known at
+the pod creation time. The container runtime may use this information to make
+preparations for all upcoming containers of the pod. E.g. setup all needed
+resources for a VM-based pod or  prepare for optimal allocation of resources of
+all the containers of the Pod. However, the container runtime may continue to
+operate as they did (before this enhancement). That is, it can safely ignore
+the per-container resource information and allocate resources for each
+container separately, one at a time, with the `CreateContainer`.
+
+```diff
+ message PodSandboxConfig {
+ 
+ ...
+ 
+     // Optional configurations specific to Linux hosts.
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++
++    // Kubernetes resource spec of the containers in the pod.
++    PodResourceConfig pod_resources = 10;
+ }
+ 
++// PodResourceConfig contains information of all resources requirements of
++// the containers of a pod.
++message PodResourceConfig {
++    repeated ContainerResourceConfig init_containers = 1;
++    repeated ContainerResourceConfig containers = 2;
++}
+ 
++// ContainerResourceConfig contains information of all resource requirements of
++// one container.
++message ContainerResourceConfig {
++    // Name of the container
++    string name= 1;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 2;
++
++    // Mounts for the container.
++    repeated Mount mounts = 3;
```

Comment:
for confidential VMs (e.g. TDX) it is needed to know all the host resources (mounts, devices specifically) that are needed to be passed through into VM before VM is created. Afterwards there is no chance to "hotplug" them. Right now creation of VM is delayed until first CreateContainer message arrives, and this creates issue if you have multiple containers in the pod.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L224-L227

>    If the hostPath doesn't exist, then runtimes should report error.

What is the expected behavior on the pod sandbox if the path doesn't exist? I could imagine a case where a subpath is initialized by an init container.

Reply:
Thanks @tallclair, good question. I think you are referring to hostPath volumes here.

I did some investigation and it looks like the CRI runtime implementations don't match the the expected behavior as stated in the CRI API (container creation should fail if the past does not exist). However in practice the CRI runtime creates the missing path (a new directory) and mounts that into the container (I don't know if there are runtime config options to affect this behavior, though). In any case, this proposal wouldn't change anything, there would be no impact on existing users.

However, for VM-based cases where mounts cannot be attached after VM-creation, the scenario you presented cannot be supported.

Some more details:

For hostPath volumes that have a `type` defined the scenario you presented is not possible – all hostPaths must exist before the sandbox is run (and before the first init container starts). The kubelet's volume manager does setup/validation for all volumes before the sandbox creation. Pod creation is pending until the checks pass. For example with `type: File`, the pod creation "get's stuck" if the file does not exist. Or, with `type: FileOrCreate` kubelet will create the file (before RunPodSandbox kicks in) if it does not exist.

However, for hostPath volumes with an empty `type` the situation is more interesting. The volume's setup/validation phase does not do any checks on the existence/type of the file. Later, kubelet just blindly hands over the mount spec to the runtime at container creation time. If some init/sidecar container happened to create the path (before the container consuming it is started) then it's mounted. If nothing creates the path before the container (consuming it) is started, then the runtime effectively does mkdir -p and mounts it. This scenario cannot be supported if mounts cannot be attached after VM-creation – the runtime would fail sandbox creation.